[STEP 2] 학습 테스트 PR 요청합니다. 문자열 계산기

badaelephant opened the pull request:
안녕하세요 근환님. 이번에는 진짜 미션이 재밌었던 것 같아요. 

처음으로, Test코드를 우선 작성하고, 실제 코드를 개발하는 것이 상당히 재밌네요!!

제가, 평일에는 밤늦게 밖에 시간이 안되서.. 늦게 보내는 것을 양해 부탁드립니다. 

좋은 하루 되세요

Commit:
refactor: Docs 작성, step구분, 테스트 소스코드 생성

Commit:
feat: splitAndSum의 기본 함수 골격 설정
1. 값이 비면 0을 리턴한다.
2. 값이 하나의 숫자면 그 숫자를 리턴한다.
3. 문자열을 스플릿하여 배열을 리턴한다.
4. 문자열 배열을 더한 값을 리턴한다.

Commit:
feat: 입력값 빈값, 숫자 검증하는 기능 구현

Commit:
feat: 문자열을 스플릿하는 메서드 구현- 커스텀 구분자 포함

Commit:
feat: 문자열을 숫자로 바꾸는 메서드 구현
1. 숫자가 아니면 RuntimeException 발생
2. 숫자가 음수면 RuntimeException 발생

Commit:
feat: stream을 활용하여 문자열의 합을 구하는 기능을 구현

Commit:
style: 스타일 적용

Commit:
docs: 구현하면서 고려한 내용 작성

Commit:
refactor: 중복되는 양수 검증이슈 수정, 패턴 스태틱 정의 수정
수정내용
1. 상수의 경우 빠진 final을 추가하여 변경 불가능 하도록, 최대한 상수표현
2. Pattern을 미리 static으로 생성하고 호출될때 사용하도록 수정
3. 기존에는 문자열 체크와 음수 체크를 나눠서 했는데, 요구조건인 양수만 받도록 수정

Commit:
docs: 참고한 블로그 링크 추가

geunhwanlee commented on src/main/java/step2/StringAddCalculator.java:
```diff
@@ -0,0 +1,57 @@
+    private static final String CUSTOM_DELIMITER_REGEX  = "//(.)\n(.*)";
+    private static final String POSITIVE_NUMBER_REGEX = "^[0-9]+$";
+    private static final String NOT_POSITIVE_NUMBER = "0~9 사이의 숫자가 아닙니다.";
+    private static final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile(CUSTOM_DELIMITER_REGEX);
```
Pattern 객체를 static 상수로 선언한 것 좋습니다 👍🏻

badaelephant replied:
감사합니다. ^--^

geunhwanlee commented on src/main/java/step2/StringAddCalculator.java:
```diff
@@ -0,0 +1,57 @@
+        if (isStringNullOrEmpty(text)) {
+            return DEFAULT_EMPTY_RETURN;
+        }
+        String[] splitNumbers = splitNumber(text);
+        validateNumbersPositive(splitNumbers);
+        return addAllNumbers(splitNumbers);
```
> 1. 문자를 구분자로 분리한다.
> 2. 분리된 문자들이 0~9 사이의 숫자인지 검증한다.
> 3. 문자들의 합을 계산한다.

StringAddCalculator가 위의 세가지 역할을 모두 하고있는데
다른 객체에게 위의 역할을 위임하는건 어떨까요?🤔

예를 들면 Number라는 객체에게 `0~9 사이의 숫자인지 검증한다`와 `Number 끼리의 합을 계산한다`는 역할을 위임할 수 있을것 같아요!

추가로 `문자를 구분자로 분리한다`는 역할은 어떤 객체에게 위임할 수 있을지도 고민해 보시면 좋을것 같습니다.

badaelephant replied:
넵 확실히 역할은 잘게 구분하면 구분할 수록 좋은 것 같습니다. 
궁금 한 것이 하나 있는데요, 지금 구조는 
SplitFactory : 문자열 구분
NumberValidator : 문자 검증
StringAddCalculator : 문자 합 계산 이렇게 각각 나뉘어져 있는데요
동일하게 클래스는 3개로 나누되

SpringCalculator는 SplitFactory라는 객체 하나를 injection 받고
SplitFactory 에서 NumberValidator라는 검증 클래스를 injection 받는 방식으로 구현하는 것은 어떨까요? 

그러면 StringAddCalculator는 2개가 아닌 1개의 클래스에만 의존하게되어 좀 더 역할과 의존관계가 명확해지지 않을까 생각이 드는데, 근환님 의견이 궁금합니다.

geunhwanlee replied:
우선 의존관계가 명확하다는건 뭘까요? 직접적으로 의존하는 객체가 적을수록 과연 의존관계가 명확할 지 생각해보는게 좋을것 같아요!
전 개인적으로 SplitFactory는 문자열을 구분자로 구분하는 역할만 하는게 더 명확하다고 생각이 드네요.😊

geunhwanlee commented on src/test/java/step2/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,48 @@
+class StringAddCalculatorTest {
+
+    @Test
+    public void splitAndSum_null_또는_빈문자() {
```
@ParameterizedTest @NullAndEmptySource 를 통해 null과 빈문자를 파라미터로 받는 건 어떨까요?

badaelephant replied:
NullAndEmptySource는 모르고 있었는데 이번 기회에 알게 되었네요. 감사합니다.

geunhwanlee commented on src/test/java/step2/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Test
+    public void splitAndSum_negative() throws Exception {
+        assertThatThrownBy(() -> StringAddCalculator.splitAndSum("-1,2,3"))
```
음수값 뿐만 아니라 숫자 외의 값(문자, 특수문자)도 테스트 해 보는건 어떨까요?

badaelephant replied:
넵, 우선 저는 테스트 코드는 미션에서 주어진 것으로 생각해서 안건들였는데, 말씀하신대로, 
0-9 사이에 해당하지 않는 모든 경우를 테스트 하는 것이 맞는것 같습니다. 피드백 감사드립니다.

geunhwanlee commented on src/main/java/step2/StringAddCalculator.java:
```diff
@@ -0,0 +1,57 @@
+    private static void validateNumbersPositive(String[] splitNumbers) {
+        boolean isValidate = Arrays.stream(splitNumbers).allMatch(number->number.matches(POSITIVE_NUMBER_REGEX));
+        if (!isValidate){
+            throw new RuntimeException(NOT_POSITIVE_NUMBER);
```
```suggestion
            throw new IllegalArgumentException(NOT_POSITIVE_NUMBER);
```
RuntimeException은 너무 포괄적인 예외인 것 같아요.

IllegalArgumentException과 같이 구체적인 예외를 던져주는것이 더 나아보입니다!

badaelephant replied:
넵 근환님, 말씀하신대로, Exception을 구체화 하는게 더 맞는것 같아서 수정하였습니다. 감사합니다.

geunhwanlee reviewed (changes requested):
안녕하세요 용석님! 🙌

2단계 미션 잘 해주셨습니다.👏

몇가지 아쉬운 점이 있어서 코멘트를 남겼어요.
한번 고민 해 보시고 궁금한 점 있으시면 언제든 질문주세요!
프리코스 과정에서 했던 경험들을 떠올리시면 도움이 될거에요.👍🏻

Commit:
fix: 피드백 내용 반영 1
문자를 구분하는 SplitFactory
숫자를 검증하는 NumberValidator
전체의 합을 담당하는 StringAddCalculator 로 역할분리
Error 형식을 IllegalArgumentException 으로 구체화

Commit:
test: 클래스 구분으로 테스트 클래스도 구분, 역할에 따라 테스트

geunhwanlee commented on src/main/java/step2/SplitFactory.java:
```diff
@@ -0,0 +1,23 @@
+    public static String[] splitNumber(String text) {
+        Matcher m = CUSTOM_DELIMITER_PATTERN.matcher(text);
+        if (m.find()) {
+            String customDelimiter = m.group(1);
```
`m.group(1)`에서 `1`이 의미하는 바가 뭘까요?
이런 케이스를 매직넘버라고 부르는데요
상수를 통해 이름을 부여할 수 있어요!
이번 단계에서는 참고만 해주세요.😊

참고하실만한 링크 보내드려요.
https://hoonmaro.tistory.com/44

geunhwanlee reviewed (approved):
리뷰사항 반영 잘 해주셨어요!

이번 단계는 여기서 머지하도록 할게요.👍🏻

그럼 다음 미션도 화이팅입니다.🔥
